### PR TITLE
Always use a template file for in process tests for much better DX

### PR DESCRIPTION
**Pull request by dawehner**
When you have test failures in subprocess tests, they can be hard to debug, as phpunit
uses a piped in php file by default. This makes it impossible for xdebug to inspect the file
and also requires phpunit to ship with a php-eval.php file.

As suggested by @sebastianbergmann in https://github.com/sebastianbergmann/phpunit/pull/2008#issuecomment-283671104
we can drop some subclassing, when we require the file parameter.

Note: I removed the windows class, as the php bug mentioned in it got solved in 2014, so it should be part of php7.

As suggested by sebastian this also moves the class, removes the factory etc.

### CODE REVIEW

**Comment by sebastianbergmann** [comment]
A lot of tests fail. Can you have a look? Thanks.

**Comment by dawehner** [comment]
Sure, I just didn't thought about running them :)

**Comment by dawehner** [comment]
I was able to fix a good amount of the failures. Let's see I can figure out the others.

File: src/Util/PHP.php [hunk 1]
```diff
@@ -63,7 +66,7 @@ public function __construct()
         $this->runtime = new Runtime();
     }
 
-    /**
+        /**
```
Comment (Bilge) [inline src/Util/PHP.php]: Why?

Comment (dawehner) [inline src/Util/PHP.php]: Thank you for the pointer! No reason of course. I fixed that.

**Comment by sebastianbergmann** [comment]
@dawehner Any progress on those remaining test failures? Thanks!

**Comment by dawehner** [comment]
@sebastianbergmann Not yet sadly. I'll try to work on it maybe tonight or more realistic on the weekend. To be honest some of them at least seem to be non trivial for me.

**Comment by sebastianbergmann** [comment]
@dawehner Thank you for the update.

**Comment by sebastianbergmann** [comment]
Sorry to bother you again: any progress? Thanks!

**Comment by dawehner** [comment]
Sorry :( I hope I have some time tomorrow or tonight to debug what is going on. I had hopes to pull some data out of sun when I saw him recently, but you know, sometimes that's hard.

**Comment by sebastianbergmann** [comment]
Any update?

**Comment by dawehner** [comment]
Nope, sorry :(

Sebastian Bergmann <notifications@github.com> schrieb am Di., 4. Juli 2017
um 10:08 Uhr:

**Comment by sebastianbergmann** [comment]
@dawehner Damn! I totally forgot to talk to you about this in Vienna.

File: src/Util/PHP.php [hunk 2]
```diff
@@ -7,25 +7,28 @@
...
 use PHPUnit\Framework\SyntheticError;
-use PHPUnit\Util\InvalidArgumentHelper;
-use SebastianBergmann\Environment\Runtime;
 
 /**
- * Utility methods for PHP sub-processes.
+ * Default utility for PHP sub-processes.
  */
-abstract class AbstractPhpProcess
+class PHP
```
Comment (keradus) [inline src/Util/PHP.php]: Name is too generic, when you give only a name to someone, he will have on idea what to expect.
for logic in this class - maybe replace most of it with symfony/process ?

Comment (sebastianbergmann) [reply src/Util/PHP.php]: See #1342. But that should not happen before #2015.

Comment (keradus) [reply src/Util/PHP.php]: :+1: for switching to process component in separated PR/issue.
Yet, still, `PHP` is way too generic and meaningless name, it shall be modified

File: src/Runner/PhptTestCase.php [hunk 3]
```diff
@@ -66,7 +66,7 @@ class PhptTestCase implements Test, SelfDescribing
      * Constructs a test case with the given filename.
      *
      * @param string             $filename
-     * @param AbstractPhpProcess $phpUtil
+     * @param \PHPUnit\Util\PHP  $phpUtil
```
Comment (keradus) [inline src/Runner/PhptTestCase.php]: changing param type for regular method is BC breaker.
@sebastianbergmann , do you treat changing `__construct` params as BC breaker ? I know some do, some don;t

File: src/Util/PHP/WindowsPhpProcess.php [hunk 4]
```diff
@@ -1,43 +0,0 @@
...
-
-/**
- * Windows utility for PHP sub-processes.
- *
- * Reading from STDOUT or STDERR hangs forever on Windows if the output is
- * too large.
- *
- * @see https://bugs.php.net/bug.php?id=51800
- */
-class WindowsPhpProcess extends DefaultPhpProcess
```
Comment (keradus) [inline src/Util/PHP/WindowsPhpProcess.php]: now my code depending on this public class will stop working :(

Comment (sebastianbergmann) [reply src/Util/PHP/WindowsPhpProcess.php]: Just because a class is `public` does not mean that is part of the public API of a package.

Comment (keradus) [reply src/Util/PHP/WindowsPhpProcess.php]: damn, without any indicator that it's not part of public API, it's hard to know that :(
so one could rely on it

Comment (sebastianbergmann) [reply src/Util/PHP/WindowsPhpProcess.php]: True. But should I really have to add `@internal` to almost all classes?

Comment (keradus) [reply src/Util/PHP/WindowsPhpProcess.php]: There are 2 ways, either add `@internal` to non-public, or `@api` to public.
Without this info, user cannot know that it's internal, so he assume it's open to use and follows SEMVER.

note: Usually, I mark most classes with `@internal` and I'm making most classes final (so non-internal got final), leaving only few of them open to extend via inheritance.

File: src/Runner/PhptTestCase.php (hunk 3 above)
Comment (dawehner) [reply src/Runner/PhptTestCase.php]: This is a good question!

**Comment by dawehner** [comment]
@sebastianbergmann 
Oh damnit indeed. I cannot promise anything to be honest, there are just so many other things.
I think in terms of debuggability it would be still a great improvement, but I understand if you would close the issue as noone actively works on it.

**Comment by sebastianbergmann** [comment]
This branch has conflicts that must be resolved, please send a new pull request when you're ready.

**Pull request closed without merging**